sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1928,33 +1925,51 @@ class Analyzer(override val catalogManager: CatalogManager)
    * returning the original attribute. In this case `d` will be resolved in subsequent passes
    * after `b` is resolved.
    */
-  protected[sql] def resolveExpressionBottomUp(
+  def resolveExpressionByPlanOutput(
       expr: Expression,
       plan: LogicalPlan,
       throws: Boolean = false): Expression = {
-    if (expr.resolved) return expr
-    // Resolve expression in one round.
-    // If throws == false or the desired attribute doesn't exist
-    // (like try to resolve `a.b` but `a` doesn't exist), fail and return the origin one.
-    // Else, throw exception.
-    try {
-      expr transformUp {
-        case GetColumnByOrdinal(ordinal, _) => plan.output(ordinal)
-        case u @ UnresolvedAttribute(nameParts) =>
-          val result =
-            withPosition(u) {
-              plan.resolve(nameParts, resolver)
-                .orElse(resolveLiteralFunction(nameParts, u, plan))
-                .getOrElse(u)
-            }
-          logDebug(s"Resolving $u to $result")
-          result
-        case UnresolvedExtractValue(child, fieldName) if child.resolved =>
-          ExtractValue(child, fieldName, resolver)
-      }
-    } catch {
-      case a: AnalysisException if !throws => expr
-    }
+    resolveExpression(
+      expr,
+      resolveColumnByName = u => {
+        plan.resolve(u.nameParts, resolver).orElse(resolveLiteralFunction(u.nameParts, u, plan))
```

Comment:
Could we move the `orElse` part into `resolveExpression `? It seems it is the same between `resolveExpressionByPlanOutput` and `resolveExpressionByPlanChildren `.